File: electronics_shop_backend/src/main/java/com/eshop/electronics/service/ProductService.java
package com.eshop.electronics.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.stereotype.Service;

import com.eshop.electronics.model.Product;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ProductService {

    @Autowired
    private MongoTemplate mongoTemplate;

    private String COLLECTION = "toys";
    private String COLLECTION = "electronics";

    public List<Product> getAllProducts() {
        return mongoTemplate.findAll(Product.class, COLLECTION);
    }

    public List<Product> getProductsByCategory(String category) {
        return getAllProducts().stream()
                .filter(p -> p.getCategory().equals(category))
                .collect(Collectors.toList());
    }

    public List<Product> searchProductsByName(String name) {
        return getAllProducts().stream()
                .filter(p -> p.getName().toLowerCase().contains(name.toLowerCase()))
                .collect(Collectors.toList());
    }

    public List<Product> getProductsSortedByPriceAsc() {
        return getAllProducts().stream()
                .sorted(Comparator.comparingDouble(Product::getPrice))
                .collect(Collectors.toList());
    }

    public List<Product> getProductsSortedByPriceDesc() {
        return getAllProducts().stream()
                .sorted(Comparator.comparingDouble(Product::getPrice).reversed())
                .collect(Collectors.toList());
    }

    public Product getProductById(String id) {
        return getAllProducts().stream()
                .filter(p -> p.getId().equals(id))
                .findFirst()
                .orElse(null);
    }
}

File: electronics_shop_frontend/src/services/productService.ts
import type { Product } from "../types";
import { apiGet } from "./api";

export const productService = {
  async getAllProducts(): Promise<Product[]> {
    return apiGet<Product[]>("/products");
  },

  async getProductById(id: string): Promise<Product> {
    return apiGet<Product>(`/products/${id}`);
  },

  async getProductsByCategory(category: string): Promise<Product[]> {
    return apiGet<Product[]>(
      `/products/category/${encodeURIComponent(category)}`
    );
  },

  async searchProducts(name: string): Promise<Product[]> {
    return apiGet<Product[]>(`/products/search/${encodeURIComponent(name)}`);
  },

  async getProductsSortedByPriceAsc(): Promise<Product[]> {
    return apiGet<Product[]>("/products/sort/price/asc");
  },

  async getProductsSortedByPriceDesc(): Promise<Product[]> {
    return apiGet<Product[]>("/products/sort/price/desc");
  },
};
